### PR TITLE
[Merged by Bors] - feat: port Order.LocallyFinite

### PR DESCRIPTION
**Pull request by dupuisf**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Order/LocallyFinite.lean [hunk 1]
```diff
@@ -0,0 +1,1380 @@
...
+attribute [local simp] Option.mem_iff
+
+private lemma aux (x : α) (p : α → Prop) :
+    (∃ a : α, p a ∧ Option.some a = Option.some x) ↔ p x := by
+  -- Porting note: `simp [Option.some_inj]` has no effect
+  constructor
+  · rintro ⟨x', hx, hx'⟩
+    obtain rfl := Option.some_inj.mp hx'
+    exact hx
+  · exact fun h => ⟨x, h, rfl⟩
```
Comment (j-loreaux) [inline Mathlib/Order/LocallyFinite.lean:1057]: Is this to work around the lack of `.match_1`? The changes here seemed rather verbose, but I haven't tried to simplify.

**Review by j-loreaux — COMMENTED** [review]
This looks okay to me aside from two issues, which may not matter:
1. The changes to the `locallyFiniteOrder` instance starting on line 1059 seem rather complicated, but I'm not sure if this is just something we need to live with.
2. The naming of `mem_Icc` instead of `mem_icc` (and similar occurrences), while in accordance with the `Set` versions, seems to me to buck the naming conventions. I have asked on [Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Naming.20convention/near/323524099)

File: Mathlib/Order/LocallyFinite.lean (hunk 1 above)
Comment (Ruben-VandeVelde) [reply Mathlib/Order/LocallyFinite.lean:1057]: Yeah. I'm not sure what's making this harder in lean4, but I'd prefer not spending too long debugging it.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**